src/framework/preferences/js/SelfVoicingEnactor.js

```diff
@@ -92,45 +108,280 @@ var fluid_3_0_0 = fluid_3_0_0 || {};
         },
         strings: {
             welcomeMsg: "text to speech enabled"
+        },
+        events: {
+            onReadFromDOM: null
+        },
+        listeners: {
+            "{tts}.events.utteranceOnEnd": {
+                listener: "fluid.prefs.enactor.selfVoicing.handleUtteranceEndEvent",
+                args: ["{that}"],
+                namespace: "cleanupSelfVoicing"
+            },
+            "{tts}.events.utteranceOnBoundary": {
+                listener: "fluid.prefs.enactor.selfVoicing.handleUtteranceBoundaryEvent",
+                args: ["{that}", "{arguments}.0"],
+                namespace: "highlightUtteranceText"
+            },
+            "onDestroy.detachRange": {
+                "this": "{that}.range",
+                method: "detach"
+            }
         }
     });
 
-    fluid.prefs.enactor.selfVoicing.handleSelfVoicing = function (welcomeMsg, queueSpeech, readFromDOM, cancel, enabled) {
+    fluid.prefs.enactor.selfVoicing.handleSelfVoicing = function (that, welcomeMsg, queueSpeech, readFromDOM, cancel, enabled) {
+        that.parseQueue = [];
         if (enabled) {
+            //TODO: It seems that we push null so that the parseQueue and queued speech remain in sync, given the
+            //      welcomeMsg pushed into the speech queue.
+            //      We should try to remove the need to push null.
+            that.parseQueue.push(null);
             queueSpeech(welcomeMsg, true);
             readFromDOM();
         } else {
             cancel();
         }
     };
 
+    fluid.prefs.enactor.selfVoicing.handleUtteranceBoundaryEvent = function (that, e) {
+        if (that.parseQueue[0]) {
+            fluid.prefs.enactor.selfVoicing.highlight(that, e.charIndex);
+        }
+    };
+
+    /**
+     * Unwraps the contents of the element by removing the tag surrounding the content and placing the content
+     * as a node within the element's parent. The parent is also normalized to combine any adjacent textnodes.
+     *
+     * @param {String|jQuery|element} elm - element to unwrap
+     */
+    fluid.prefs.enactor.selfVoicing.unWrap = function (elm) {
+        elm = $(elm);
+
+        // remove previous marks and normalize parent to clean up textnodes
+        if (elm.length) {
+            var parent = elm.parent();
+            elm.contents().unwrap();
+            parent[0].normalize();
+        }
+    };
+
+    fluid.prefs.enactor.selfVoicing.handleUtteranceEndEvent = function (that) {
+        that.parseQueue.shift();
+        that.parseIndex = 0;
+        fluid.prefs.enactor.selfVoicing.unWrap(that.locate("mark"));
```

Comment:
Very awkward mix of "model-based" action on the queues and dispatch to a global function. The last action should be implemented by a suitably-named invoker, and this handler should be split into two separate listeners